File: backend/services/pfz_service.py
"""
ORCA — Potential Fishing Zones (PFZ) Service (Backend)

Integrates verified official INCOIS (Indian National Centre for Ocean Information Services)
WebGIS OGC WFS endpoints:
  1. PFZ Lines: https://www.incois.gov.in/geoserver/PFZ_Automation/ows
     (Layer: PFZ_Automation:pfzlines)
  2. PFZ Landing Centre Advisories: https://www.incois.gov.in/geoserver/PFZ_LandingCentres/ows
     (Layer: PFZ_LandingCentres:LandingCenters_29Apr2024)
  3. PFZ Coastal Sectors: https://www.incois.gov.in/geoserver/PFZ_Sectors/ows
     (Layer: PFZ_Sectors:sector_new)

Provides live, dynamic, location-specific PFZ intelligence with full provenance.
DO NOT fabricate or hardcode PFZ coordinates.
"""

import math
import time
from datetime import datetime, timezone
import httpx
from typing import Dict, Any, List, Optional, Tuple

# Official INCOIS GeoServer WFS Base Endpoints
INCOIS_PFZ_LINES_WFS = (
    "https://www.incois.gov.in/geoserver/PFZ_Automation/ows"
    "?service=WFS&version=1.1.0&request=GetFeature"
    "&typeName=PFZ_Automation:pfzlines&outputFormat=application/json"
)

INCOIS_LANDING_CENTRES_WFS = (
    "https://www.incois.gov.in/geoserver/PFZ_LandingCentres/ows"
    "?service=WFS&version=1.1.0&request=GetFeature"
    "&typeName=PFZ_LandingCentres:LandingCenters_29Apr2024&outputFormat=application/json"
)

# In-memory caching with 15-minute TTL to reduce upstream server load
_CACHE: Dict[str, Any] = {}
_CACHE_TTL_SECONDS = 900  # 15 minutes


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculates geodesic distance between two points in kilometers."""
    R = 6371.0
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (
        math.sin(dlat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(dlon / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return R * c


def calculate_bearing(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculates compass bearing from point 1 to point 2 in degrees."""
    dlon = math.radians(lon2 - lon1)
    y = math.sin(dlon) * math.cos(math.radians(lat2))
    x = math.cos(math.radians(lat1)) * math.sin(math.radians(lat2)) - math.sin(
        math.radians(lat1)
    ) * math.cos(math.radians(lat2)) * math.cos(dlon)
    initial_bearing = math.degrees(math.atan2(y, x))
    return (initial_bearing + 360) % 360


def degrees_to_compass_dir(degrees: float) -> str:
    """Converts bearing degrees to 16-point compass direction."""
    dirs = [
        "N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
        "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW",
    ]
    idx = round(degrees / (360 / len(dirs))) % len(dirs)
    return dirs[idx]


def evaluate_validity(validity_str: Optional[str]) -> Tuple[bool, str, str]:
    """
    Evaluates whether an INCOIS validity date is currently active or historical.
    Returns (is_currently_valid, status_code, formatted_label).
    """
    if not validity_str:
        return False, "UNKNOWN", "Validity Not Specified"
    try:
        val_clean = validity_str.replace("Z", "+00:00")
        val_dt = datetime.fromisoformat(val_clean)
        now_dt = datetime.now(timezone.utc)
        is_valid = val_dt >= now_dt
        formatted_date = val_dt.strftime("%d-%b-%Y")
        if is_valid:
            return True, "CURRENT", f"{formatted_date} (Current Advisory)"
        else:
            return False, "HISTORICAL_ARCHIVE", f"{formatted_date} (Historical INCOIS Advisory Cycle)"
    except Exception:
        return False, "UNPARSED", validity_str


async def fetch_incois_raw_wfs() -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
    """
    Fetches raw GeoJSON features from INCOIS WFS endpoints with caching.
    Returns (pfz_lines_features, landing_centre_features).
    """
    now = time.time()
    cached_time = _CACHE.get("timestamp", 0)
    if (now - cached_time < _CACHE_TTL_SECONDS) and "lines" in _CACHE and "landing_centres" in _CACHE:
        return _CACHE["lines"], _CACHE["landing_centres"]

    headers = {
        "User-Agent": "ORCA Marine Intelligence Agent (ISRO SIH 26176 / INCOIS Integration)"
    }

    lines_features = []
    lc_features = []

    async with httpx.AsyncClient(timeout=25.0, verify=False, follow_redirects=True, headers=headers) as client:
        # Fetch PFZ Lines
        try:
            res_lines = await client.get(INCOIS_PFZ_LINES_WFS)
            if res_lines.status_code == 200:
                lines_data = res_lines.json()
                lines_features = lines_data.get("features", [])
        except Exception as e:
            print(f"[PFZ Service] Warning: Failed to fetch INCOIS PFZ Lines: {e}")

        # Fetch Landing Centres
        try:
            res_lc = await client.get(INCOIS_LANDING_CENTRES_WFS)
            if res_lc.status_code == 200:
                lc_data = res_lc.json()
                lc_features = lc_data.get("features", [])
        except Exception as e:
            print(f"[PFZ Service] Warning: Failed to fetch INCOIS Landing Centres: {e}")

    # Update cache if we got data
    if lines_features or lc_features:
        _CACHE["lines"] = lines_features
        _CACHE["landing_centres"] = lc_features
        _CACHE["timestamp"] = now

    return lines_features, lc_features


async def get_pfz_assessment(lat: float, lon: float, max_radius_km: float = 250.0) -> Dict[str, Any]:
    """
    Evaluates real INCOIS Potential Fishing Zones (PFZ) for a given marine/coastal coordinate.

    Returns:
      - nearest_advisory: Closest Landing Centre with official distance, bearing, depth range
      - active_advisories: Nearby landing centres with PFZ advisories (STATUS='YES')
      - pfz_lines: Real MultiLineString geometries of PFZ thermal/chlorophyll fronts
      - advisory_metadata: Exact validity dates, freshness status, dataset update timestamps
    """
    lines_features, lc_features = await fetch_incois_raw_wfs()

    # 1. Process Landing Centres near the requested coordinate
    active_advisories = []
    for feat in lc_features:
        props = feat.get("properties", {})
        status = props.get("STATUS", "").strip().upper()
        if status != "YES":
            continue

        lc_lat = props.get("LATITUDE")
        lc_lon = props.get("LONGITUDE")
        if lc_lat is None or lc_lon is None:
            continue

        dist_from_loc = haversine_km(lat, lon, float(lc_lat), float(lc_lon))
        if dist_from_loc <= max_radius_km:
            bearing = props.get("BEARING")
            direction = props.get("DIRECTION") or (degrees_to_compass_dir(bearing) if bearing is not None else "N/A")

            # Target PFZ coordinate DMS
            lat_dms = f"{props.get('LATITUDE_D', '')}°{props.get('LATITUDE_M', '')}'{props.get('LATITUDE_S', '')}\"{props.get('LATITUDE_1', 'N')}"
            lon_dms = f"{props.get('LONGITUDE_', '')}°{props.get('LONGITUDE1', '')}'{props.get('LONGITUD_1', '')}\"{props.get('LONGITUD_2', 'E')}"

            # Evaluate advisory validity date freshness
            is_valid, val_status, val_formatted = evaluate_validity(props.get("VALIDITY_D"))

            active_advisories.append({
                "id": f"incois-lc-{props.get('OBJECTID', '')}-{props.get('LC_UNIQUE_', '')}",
                "landing_center": props.get("LC_NAME", "Unknown"),
                "district": props.get("DIST_NAME", ""),
                "sector": props.get("SECTOR_NAM", ""),
                "lc_coordinates": {"latitude": float(lc_lat), "longitude": float(lc_lon)},
                "distance_from_query_km": round(dist_from_loc, 1),
                "advisory_distance_from_km": props.get("DISTANCE_F"),
                "advisory_distance_to_km": props.get("DISTANCE_T"),
                "direction": direction,
                "bearing_degrees": bearing,
                "depth_from_m": props.get("DEPTH_FROM"),
                "depth_to_m": props.get("DEPTH_TO"),
                "target_dms": {"latitude": lat_dms, "longitude": lon_dms},
                "forecast_date": props.get("FORECAST_D"),
                "validity_date": props.get("VALIDITY_D"),
                "validity_formatted": val_formatted,
                "is_currently_valid": is_valid,
                "validity_status": val_status,
                "updated_date": props.get("UPDATED_DA"),
                "dataset_updated": "29-Apr-2024 (INCOIS GeoServer Layer)",
                "forecast_issue_id": props.get("FORECAST_I"),
                "status": "OFFICIAL_ADVISORY",
                "source": "INCOIS — Official PFZ Advisory",
            })

    # Sort by distance from current vessel/assessment position
    active_advisories.sort(key=lambda x: x["distance_from_query_km"])

    # 2. Process ALL real INCOIS PFZ lines (nationwide) and compute distance to query location
    nationwide_lines = []
    relevant_lines = []
    for feat in lines_features:
        props = feat.get("properties", {})
        geom = feat.get("geometry", {})
        coords = geom.get("coordinates", [])

        line_min_dist = float("inf")
        if geom.get("type") == "MultiLineString":
            for line in coords:
                for pt in line:
                    d = haversine_km(lat, lon, pt[1], pt[0])
                    if d < line_min_dist:
                        line_min_dist = d
        elif geom.get("type") == "LineString":
            for pt in coords:
                d = haversine_km(lat, lon, pt[1], pt[0])
                if d < line_min_dist:
                    line_min_dist = d

        line_obj = {
            "uid": props.get("UID", ""),
            "state_name": props.get("State_Name", ""),
            "category": props.get("Category", "sst_ocean_color"),
            "julian_day": props.get("Julian_day"),
            "year": props.get("Year"),
            "length_km": round(float(props.get("Length", 0)), 1) if props.get("Length") else None,
            "distance_km": round(line_min_dist, 1) if line_min_dist != float("inf") else 0.0,
            "geometry": geom,
            "source": "INCOIS — Official PFZ Lines WFS",
        }

        nationwide_lines.append(line_obj)
        if line_min_dist <= (max_radius_km + 150.0):
            relevant_lines.append(line_obj)

    relevant_lines.sort(key=lambda x: x["distance_km"])
    nationwide_lines.sort(key=lambda x: x["distance_km"])

    # Determine overall status and freshness
    nearest_advisory = active_advisories[0] if active_advisories else None
    has_pfz_data = len(nationwide_lines) > 0 or nearest_advisory is not None
    is_currently_valid = nearest_advisory["is_currently_valid"] if nearest_advisory else False
    advisory_available = nearest_advisory is not None

    if advisory_available:
        advisory_message = f"Official INCOIS landing-centre advisory available for {nearest_advisory['landing_center']} sector."
    else:
        advisory_message = "No current location-specific landing-centre advisory found for this sector; nationwide satellite PFZ frontal lines remain active."

    return {
        "available": has_pfz_data,
        "advisory_available": advisory_available,
        "advisory_message": advisory_message,
        "is_currently_valid": is_currently_valid,
        "source": "INCOIS",
        "source_type": "Official PFZ Advisory (WebGIS GeoServer)",
        "query_coordinates": {"latitude": lat, "longitude": lon},
        "search_radius_km": max_radius_km,
        "nearest_advisory": nearest_advisory,
        "total_active_advisories_found": len(active_advisories),
        "active_advisories": active_advisories[:20],  # Top 20 nearby
        "total_pfz_lines_found": len(relevant_lines),
        "pfz_lines": relevant_lines[:15],  # Regional closest lines
        "nationwide_pfz_lines": nationwide_lines,  # All nationwide INCOIS PFZ lines for map rendering
        "total_nationwide_lines": len(nationwide_lines),
        "advisory_metadata": {
            "authority": "Indian National Centre for Ocean Information Services (INCOIS)",
            "ministry": "Ministry of Earth Sciences, Govt. of India",
            "service_type": "Multi-Mission Satellite Ocean Color & SST Thermal Fronts",
            "dataset_updated": "29-Apr-2024",
            "dataset_layer": "LandingCenters_29Apr2024 / PFZ_Automation:pfzlines",
            "is_currently_valid": is_currently_valid,
            "validity_date": nearest_advisory.get("validity_date") if nearest_advisory else None,
            "validity_formatted": nearest_advisory.get("validity_formatted") if nearest_advisory else "Reference dataset / No sector bulletin",
            "validity_status": nearest_advisory.get("validity_status") if nearest_advisory else "NO_SECTOR_ADVISORY",
            "pfz_lines_run": f"Julian Day {nationwide_lines[0]['julian_day']}, Year {nationwide_lines[0]['year']}" if nationwide_lines else None,
        },
        "provenance_note": (
            "PFZ data retrieved directly from official INCOIS GeoServer OGC WFS services. "
            "Landing centre advisory vector records are from the official INCOIS 28-Apr-2024 bulletin cycle. "
            "Automated PFZ frontal lines are from INCOIS PFZ_Automation. "
            "ORCA strictly distinguishes official government records from current validity and never fabricates dates."
        ),
    }

File: src/services/pfzService.ts
/**
 * ORCA — Potential Fishing Zone (PFZ) Frontend Service
 * 
 * Communicates with ORCA FastAPI backend to retrieve verified official INCOIS
 * Potential Fishing Zones (PFZ), Landing Centre Advisories, and Satellite PFZ Lines.
 * 
 * Provenance: Indian National Centre for Ocean Information Services (INCOIS)
 * Ministry of Earth Sciences, Government of India.
 */

import { FASTAPI_BASE_URL } from '@/config/api';

export interface PFZCoordinates {
  latitude: number;
  longitude: number;
}

export interface PFZAdvisory {
  id: string;
  landing_center: string;
  district?: string;
  sector?: string;
  lc_coordinates: PFZCoordinates;
  distance_from_query_km: number;
  advisory_distance_from_km?: number | null;
  advisory_distance_to_km?: number | null;
  direction?: string;
  bearing_degrees?: number | null;
  depth_from_m?: number | null;
  depth_to_m?: number | null;
  target_dms?: {
    latitude: string;
    longitude: string;
  };
  forecast_date?: string | null;
  validity_date?: string | null;
  validity_formatted?: string | null;
  is_currently_valid?: boolean;
  validity_status?: string;
  updated_date?: string | null;
  dataset_updated?: string;
  forecast_issue_id?: string | number | null;
  status: string;
  source: string;
}

export interface PFZLineFeature {
  uid: string;
  state_name?: string;
  category?: string;
  julian_day?: number | null;
  year?: number | null;
  length_km?: number | null;
  distance_km: number;
  geometry: {
    type: 'MultiLineString' | 'LineString';
    coordinates: number[][][] | number[][];
  };
  source: string;
}

export interface PFZAssessmentResponse {
  available: boolean;
  advisory_available?: boolean;
  advisory_message?: string;
  is_currently_valid: boolean;
  source: string;
  source_type: string;
  query_coordinates: PFZCoordinates;
  search_radius_km: number;
  nearest_advisory: PFZAdvisory | null;
  total_active_advisories_found: number;
  active_advisories: PFZAdvisory[];
  total_pfz_lines_found: number;
  pfz_lines: PFZLineFeature[];
  regional_pfz_lines?: PFZLineFeature[];
  total_regional_lines?: number;
  nationwide_pfz_lines?: PFZLineFeature[];
  total_nationwide_lines?: number;
  advisory_metadata: {
    authority: string;
    ministry: string;
    service_type: string;
    dataset_updated?: string;
    dataset_layer?: string;
    is_currently_valid?: boolean;
    localized_advisory_status?: string;
    regional_vectors_status?: string;
    validity_date?: string | null;
    validity_formatted?: string | null;
    validity_status?: string;
    pfz_lines_run?: string | null;
  };
  provenance_note: string;
}

/**
 * Fetches real INCOIS Potential Fishing Zones (PFZ) assessment from the ORCA backend.
 */
export async function fetchFastAPIPFZ(
  lat: number,
  lon: number,
  radiusKm: number = 250
): Promise<PFZAssessmentResponse> {
  try {
    const res = await fetch(
      `${FASTAPI_BASE_URL}/api/pfz?lat=${encodeURIComponent(lat)}&lon=${encodeURIComponent(lon)}&radius_km=${encodeURIComponent(radiusKm)}`,
      {
        headers: { 'Accept': 'application/json' },
        cache: 'no-store',
      }
    );

    if (!res.ok) {
      throw new Error(`PFZ service HTTP ${res.status}: ${res.statusText}`);
    }

    const data: PFZAssessmentResponse = await res.json();
    return data;
  } catch (err) {
    console.warn('[PFZ Service] Unable to reach backend PFZ endpoint:', err);
    return {
      available: false,
      advisory_available: false,
      advisory_message: 'PFZ data temporarily unavailable. Verify backend connection.',
      is_currently_valid: false,
      source: 'INCOIS',
      source_type: 'Official PFZ Advisory',
      query_coordinates: { latitude: lat, longitude: lon },
      search_radius_km: radiusKm,
      nearest_advisory: null,
      total_active_advisories_found: 0,
      active_advisories: [],
      total_pfz_lines_found: 0,
      pfz_lines: [],
      nationwide_pfz_lines: [],
      total_nationwide_lines: 0,
      advisory_metadata: {
        authority: 'Indian National Centre for Ocean Information Services (INCOIS)',
        ministry: 'Ministry of Earth Sciences, Govt. of India',
        service_type: 'Multi-Mission Satellite Ocean Color & SST Thermal Fronts',
        dataset_updated: 'Unavailable',
        is_currently_valid: false,
        validity_formatted: 'Connection Unavailable',
        validity_status: 'UNAVAILABLE',
      },
      provenance_note: 'Backend PFZ service currently unreachable. Verify FastAPI server status.',
    };
  }
}
